src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -59,76 +61,321 @@ G1ConcurrentRefineThread* G1ConcurrentRefineThreadControl::create_refinement_thr
+  }
+}
+
+void G1ConcurrentRefineThreadControl::worker_threads_do(ThreadClosure* tc) {
+  if (is_refinement_enabled()) {
+    _workers->threads_do(tc);
+  }
+}
+
+void G1ConcurrentRefineThreadControl::stop() {
+  if (is_refinement_enabled()) {
+    _control_thread->stop();
+  }
+}
+
+G1ConcurrentRefineSweepState::G1ConcurrentRefineSweepState(uint max_reserved_regions) :
+  _state(State::Idle),
+  _sweep_start_epoch(0),
+  _sweep_table(new G1CardTableClaimTable(G1CollectedHeap::get_chunks_per_region_for_merge())),
+  _stats()
+{
+  _sweep_table->initialize(max_reserved_regions);
+}
+
+G1ConcurrentRefineSweepState::~G1ConcurrentRefineSweepState() {
+  delete _sweep_table;
+}
+
+void G1ConcurrentRefineSweepState::set_state_start_time() {
+  _state_start[static_cast<uint>(_state)] = Ticks::now();
+}
+
+Tickspan G1ConcurrentRefineSweepState::get_duration(State start, State end) {
+  return _state_start[static_cast<uint>(end)] - _state_start[static_cast<uint>(start)];
+}
+
+void G1ConcurrentRefineSweepState::reset_stats() {
+  stats()->reset();
+}
+
+void G1ConcurrentRefineSweepState::add_yield_duration(jlong duration) {
+  stats()->inc_yield_duration(duration);
+}
+
+size_t G1ConcurrentRefineSweepState::refinement_epoch() {
+  return G1CollectedHeap::heap()->refinement_epoch();
+}
+
+bool G1ConcurrentRefineSweepState::in_sweep_epoch() const {
+  return _sweep_start_epoch == refinement_epoch();
+}
+
+bool G1ConcurrentRefineSweepState::advance_state(State next_state) {
+  bool result = in_sweep_epoch();
+  if (result) {
+    _state = next_state;
+  } else {
+    _state = State::Idle;
+  }
+  return result;
+}
+
+void G1ConcurrentRefineSweepState::assert_state(State expected) {
+  assert(_state == expected, "must be %s but is %s", state_name(expected), state_name(_state));
+}
+
+void G1ConcurrentRefineSweepState::start_work() {
+  assert_state(State::Idle);
+
+  set_state_start_time();
+
+  _sweep_start_epoch = refinement_epoch();
+
+  _stats.reset();
+
+  advance_state(State::SwapGlobalCT);
+}
+
+bool G1ConcurrentRefineSweepState::swap_global_card_table() {
+  assert_state(State::SwapGlobalCT);
+
+  set_state_start_time();
+
+  {
+    // We can't have any new threads being in the process of created while we
+    // swap the card table because we read the current card table state during
+    // initialization.
+    // A safepoint may occur during that time, so leave the STS temporarily.
+    SuspendibleThreadSetLeaver sts_leave;
+
+    MutexLocker mu(Threads_lock);
+    // A GC that advanced the epoch might have happened, which already switched
+    // The global card table. Do nothing.
+    if (in_sweep_epoch()) {
+      G1BarrierSet::g1_barrier_set()->swap_global_card_table();
     }
+  }
+
+  return advance_state(State::SwapJavaThreadsCT);
+}
+
+bool G1ConcurrentRefineSweepState::swap_java_threads_ct() {
+  assert_state(State::SwapJavaThreadsCT);
+
+  set_state_start_time();
+
+  {
+    SuspendibleThreadSetLeaver sts_leave;
```

Comment:
Can you add some comment on why leaving the set is required? It's not obvious to me why. I'd expect handshake to work out of the box...